File: src/main/java/io/xol/chunkstories/api/item/ItemVoxel.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.item;

import io.xol.chunkstories.api.Location;
import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.entity.Controller;
import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.entity.traits.TraitVoxelSelection;
import io.xol.chunkstories.api.entity.traits.serializable.TraitCreativeMode;
import io.xol.chunkstories.api.events.player.voxel.PlayerVoxelModificationEvent;
import io.xol.chunkstories.api.events.voxel.WorldModificationCause;
import io.xol.chunkstories.api.exceptions.world.WorldException;
import io.xol.chunkstories.api.input.Input;
import io.xol.chunkstories.api.item.inventory.ItemPile;
import io.xol.chunkstories.api.player.Player;
import io.xol.chunkstories.api.sound.SoundSource.Mode;
import io.xol.chunkstories.api.voxel.Voxel;
import io.xol.chunkstories.api.world.WorldMaster;
import io.xol.chunkstories.api.world.cell.CellData;
import io.xol.chunkstories.api.world.cell.FutureCell;

import javax.annotation.Nullable;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/**
 * An item that contains voxels
 */
public class ItemVoxel extends Item implements WorldModificationCause {
    private final Content.Voxels store;

    @Nullable
    public Voxel voxel = null;
    public int voxelMeta = 0;

    public ItemVoxel(ItemDeclaration type) {
        super(type);
        store = type.store().parent().voxels();
    }

    @Override
    public String getTextureName(ItemPile pile) {
        // ItemDataVoxel idv = (ItemDataVoxel) pile.data;
        if (voxel != null)
            return "./items/icons/" + voxel.getName() + ".png";
        return "./items/icons/notex.png";
    }

    @Nullable
    public Voxel getVoxel() {
        return voxel;
        // ((ItemDataVoxel) pile.getData()).voxel;
    }

    public int getVoxelMeta() {
        return voxelMeta;
        // ((ItemDataVoxel) pile.getData()).voxelMeta;
    }

    @Override
    public boolean onControllerInput(Entity entity, ItemPile pile, Input input, Controller controller) {
        try {
            if (entity.getWorld() instanceof WorldMaster && input.getName().equals("mouse.right")) {
                // Require entities to be of the right kind
                if (!(entity instanceof WorldModificationCause)) {
                    return true;
                }

                WorldModificationCause modifierEntity = (WorldModificationCause) entity;

                boolean isEntityCreativeMode = entity.getTraits().tryWithBoolean(TraitCreativeMode.class, ecm -> ecm.get());

                Location blockLocation = entity.getTraits().tryWith(TraitVoxelSelection.class, tvs -> tvs.getBlockLookingAt(false, true));

                if (blockLocation != null) {
                    FutureCell fvc = new FutureCell(entity.getWorld().peekSafely(blockLocation));
                    fvc.setVoxel(voxel);

                    // Opaque blocks overwrite the original light with zero.
                    if (voxel.getOpaque()) {
                        fvc.setBlocklight(0);
                        fvc.setSunlight(0);
                    }

                    // Glowy stuff should glow
                    // if(voxel.getDeclaration().getEmittedLightLevel() > 0)
                    fvc.setBlocklight(voxel.getEmittedLightLevel(fvc));

                    // Player events mod
                    if (controller instanceof Player) {
                        Player player = (Player) controller;
                        CellData ctx = entity.getWorld().peek(blockLocation);
                        PlayerVoxelModificationEvent event = new PlayerVoxelModificationEvent(ctx, fvc,
                                isEntityCreativeMode ? TraitCreativeMode.CREATIVE_MODE : this, player);

                        // Anyone has objections ?
                        entity.getWorld().getGameContext().getPluginManager().fireEvent(event);

                        if (event.isCancelled())
                            return true;

                        entity.getWorld().getSoundManager().playSoundEffect("sounds/gameplay/voxel_place.ogg", Mode.NORMAL, fvc.getLocation(), 1.0f, 1.0f);
                    }

                    entity.getWorld().poke(fvc, modifierEntity);

                    // Decrease stack size
                    if (!isEntityCreativeMode) {
                        int currentAmount = pile.getAmount();
                        currentAmount--;
                        pile.setAmount(currentAmount);
                    }
                } else {
                    // No space found :/
                    return true;
                }
            }

        } catch (WorldException e) {

        }

        return false;

    }

    @Override
    public void load(DataInputStream stream) throws IOException {
        voxel = store.getVoxel(stream.readUTF()); // store.getVoxelById(stream.readInt());
        voxelMeta = stream.readByte();
    }

    @Override
    public void save(DataOutputStream stream) throws IOException {
        if (voxel == null) {
            stream.writeUTF("air");
        } else {
            stream.writeUTF(voxel.getName());
        }
        stream.writeByte(voxelMeta);
    }

    @Override
    /** Two ItemVoxel can merge if they represent the same voxel & they share the
     * same 8 bits of metadata */
    public boolean canStackWith(Item item) {
        if (item instanceof ItemVoxel) {
            ItemVoxel itemVoxel = (ItemVoxel) item;
            return super.canStackWith(itemVoxel) && itemVoxel.getVoxel().sameKind(getVoxel()) && itemVoxel.getVoxelMeta() == this.getVoxelMeta();
        }
        return false;
    }

    @Override
    public String getName() {
        if (voxel != null) {
            return voxel.getName();
        }
        return "novoxel!";
    }
}

File: src/main/java/xyz/chunkstories/api/item/ItemVoxel.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.item

import org.joml.Matrix4f
import org.joml.Vector3i
import org.joml.Vector4f
import xyz.chunkstories.api.content.Content
import xyz.chunkstories.api.content.json.asString
import xyz.chunkstories.api.entity.Controller
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.traits.TraitSight
import xyz.chunkstories.api.entity.traits.serializable.TraitCreativeMode
import xyz.chunkstories.api.events.player.voxel.PlayerVoxelModificationEvent
import xyz.chunkstories.api.events.voxel.WorldModificationCause
import xyz.chunkstories.api.exceptions.world.WorldException
import xyz.chunkstories.api.graphics.MeshMaterial
import xyz.chunkstories.api.graphics.representation.ModelInstance
import xyz.chunkstories.api.graphics.representation.ModelPosition
import xyz.chunkstories.api.graphics.representation.PointLight
import xyz.chunkstories.api.graphics.systems.dispatching.RepresentationsGobbler
import xyz.chunkstories.api.input.Input
import xyz.chunkstories.api.item.inventory.ItemPile
import xyz.chunkstories.api.physics.RayQuery
import xyz.chunkstories.api.physics.RayResult
import xyz.chunkstories.api.physics.trace
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.sound.SoundSource.Mode
import xyz.chunkstories.api.util.kotlin.toVec3d
import xyz.chunkstories.api.util.kotlin.toVec3f
import xyz.chunkstories.api.util.kotlin.toVec3i
import xyz.chunkstories.api.voxel.Voxel
import xyz.chunkstories.api.voxel.VoxelSide
import xyz.chunkstories.api.world.WorldMaster
import xyz.chunkstories.api.world.cell.Cell
import xyz.chunkstories.api.world.cell.DummyCell
import xyz.chunkstories.api.world.cell.FutureCell

/** An item that contains voxels  */
open class ItemVoxel(definition: ItemDefinition) : Item(definition), WorldModificationCause {
    private val store: Content.Voxels = definition.store.parent.voxels

    val voxel: Voxel

    init {
        voxel = store.getVoxel(definition["voxel"].asString!!)!!
    }

    override fun getTextureName(): String {
        return "voxels/textures/" + voxel.getVoxelTexture(DummyCell(0, 0, 0, voxel, 0, 0, 15), VoxelSide.FRONT).name + ".png"
    }

    override fun buildRepresentation(worldPosition: Matrix4f, representationsGobbler: RepresentationsGobbler) {
        //val customMaterial = MeshMaterial("cubeMaterial", mapOf("albedoTexture" to getTextureName(pile)))
        val customMaterials = VoxelSide.values().map { side ->
            val textureName = "voxels/textures/" + voxel.getVoxelTexture(DummyCell(0, 0, 0, voxel, 0, 0, 15), side).name + ".png"
            val material = MeshMaterial("cubeMaterial$side", mapOf("albedoTexture" to textureName))
            Pair(side.ordinal, material)
        }.toMap()


        val representation = ModelInstance(store.parent.models["voxels/blockmodels/cube/cube.dae"], ModelPosition(worldPosition).apply {
            matrix.scale(0.35f)
            matrix.translate(-0.5f, -0.0f, -0.5f)
        }, customMaterials)
        representationsGobbler.acceptRepresentation(representation, -1)

        val position = Vector4f(0f, 0f, 0f, 1f)
        worldPosition.transform(position)
        if (voxel.emittedLightLevel > 0) {
            val light = PointLight(position.toVec3f().toVec3d().add(0.0, 0.5, 0.0), voxel.voxelTextures[0].color.toVec3f().toVec3d().mul(voxel.emittedLightLevel.toDouble()))
            representationsGobbler.acceptRepresentation(light)
        }
    }

    override fun onControllerInput(entity: Entity, pile: ItemPile, input: Input, controller: Controller): Boolean {
        try {
            if (entity.world is WorldMaster && input.name == "mouse.right") {
                // Require entities to be of the right kind
                if (entity !is WorldModificationCause) {
                    return true
                }

                val modifierEntity = entity as WorldModificationCause

                val isEntityCreativeMode = entity.traits[TraitCreativeMode::class]?.enabled ?: false

                val hit: RayResult.Hit.VoxelHit = entity.traits[TraitSight::class]?.let {
                    val reach = 5.0
                    val notMe = {other: Entity -> other != entity}
                    val query = RayQuery(it.headLocation, it.lookingAt, 0.0, reach, { it.voxel.solid }, notMe)
                    when(val hit = query.trace()) {
                        is RayResult.Hit.VoxelHit -> {
                            return@let hit as RayResult.Hit.VoxelHit // Kotlin is buggy: if I remove this cast it fails
                        }
                    }

                    return@let null
                } ?: return false

                val pointedCell = hit.cell

                val normali = hit.normal.toVec3i()
                val x = hit.cell.x + normali.x
                val y = hit.cell.y + normali.y
                val z = hit.cell.z + normali.z
                val side =  when(normali) {
                    Vector3i(1, 0, 0) -> VoxelSide.RIGHT
                    Vector3i(-1, 0, 0) -> VoxelSide.LEFT
                    Vector3i(0, 1, 0) -> VoxelSide.TOP
                    Vector3i(0, -1, 0) -> VoxelSide.BOTTOM
                    Vector3i(0, 0, 1) -> VoxelSide.FRONT
                    Vector3i(0 ,0, -1) -> VoxelSide.BACK
                    else -> throw Exception("How did we get here ?")
                }

                val adjacentCell = if(y in 0 until hit.cell.world.maxHeight) {
                    hit.cell.world.peek(x, y, z)
                } else {
                    return false
                }

                if (adjacentCell.voxel.isAir() /** TODO handle overwriting blocks cases */) {
                    val futureCell = FutureCell(adjacentCell)

                    // Let's stop if this returns false
                    if (!prepareNewBlockData(futureCell, pointedCell, side, entity, hit))
                        return true

                    // Player events mod
                    if (controller is Player) {
                        //val ctx = entity.world.tryPeek(freeCell.location)
                        val event = PlayerVoxelModificationEvent(adjacentCell, futureCell,
                                if (isEntityCreativeMode) TraitCreativeMode.CREATIVE_MODE else this, controller)

                        // Anyone has objections ?
                        entity.world.gameContext.pluginManager.fireEvent(event)

                        if (event.isCancelled)
                            return true

                        entity.world.soundManager.playSoundEffect("sounds/gameplay/voxel_place.ogg", Mode.NORMAL, futureCell.location, 1.0f, 1.0f)
                    }

                    entity.world.poke(futureCell, modifierEntity)

                    // Decrease stack size
                    if (!isEntityCreativeMode) {
                        pile.amount--
                    }
                } else {
                    // No space found :/
                    return true
                }
            }

        } catch (e: WorldException) {

        }

        return false

    }

    open fun prepareNewBlockData(cell: FutureCell, adjacentCell: Cell, adjacentCellSide: VoxelSide, placingEntity: Entity, hit: RayResult.Hit.VoxelHit): Boolean {
        cell.voxel = voxel
        cell.metaData = 0

        // Opaque blocks overwrite the original light with zero.
        if (voxel.opaque) {
            cell.blocklight = 0
            cell.sunlight = 0
        }

        // Glowy stuff should glow
        cell.blocklight = voxel.getEmittedLightLevel(cell)

        return true
    }
}
